STModCellular.cpp

```diff
@@ -18,53 +18,104 @@
 #include "STModCellular.h"
 #include "mbed_wait_api.h"
 #include "mbed_trace.h"
+#include "DigitalIn.h"
 
 #define TRACE_GROUP "CELL"
 
 using namespace mbed;
 
-STModCellular::STModCellular(FileHandle *fh) : QUECTEL_UG96(fh),
+STModCellular::STModCellular(FileHandle *fh) : STMOD_CELLULAR_MODEM(fh),
     m_powerkey(MBED_CONF_STMOD_CELLULAR_POWER),
     m_reset(MBED_CONF_STMOD_CELLULAR_RESET),
     m_simsel0(MBED_CONF_STMOD_CELLULAR_SIMSEL0),
     m_simsel1(MBED_CONF_STMOD_CELLULAR_SIMSEL1),
     m_mdmdtr(MBED_CONF_STMOD_CELLULAR_MDMDTR)
 {
-    // start with modem disabled
-    m_mdmdtr.write(0);
-    m_simsel0.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION);
-    m_simsel1.write(0);
 
+    tr_debug("STModCellular creation\r\n");
+
+    /* Ensure PIN SIMs are set as input */
+    DigitalIn sim_reset(MBED_CONF_STMOD_CELLULAR_SIM_RESET);
+    DigitalIn sim_clk(MBED_CONF_STMOD_CELLULAR_SIM_CLK);
+    DigitalIn sim_data(MBED_CONF_STMOD_CELLULAR_SIM_DATA);
+
+    // start with modem disabled
+    m_powerkey.write(0);
     m_reset.write(1);
-    wait_ms(250);
+    wait_ms(200);
     m_reset.write(0);
+    wait_ms(150);
+
+    wait_ms(50);
+    m_simsel0.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION & 0x01);
+    m_simsel1.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION & 0x02);
+    wait_ms(50);
 }
 
 STModCellular::~STModCellular()
 {
 }
 
 nsapi_error_t STModCellular::soft_power_on() {
-    nsapi_error_t err = QUECTEL_UG96::soft_power_on();
+    tr_debug("STMOD cellular modem power ON\r\n");
+
+#if (MBED_CONF_STMOD_CELLULAR_TYPE == STMOD_UG96)
+    tr_debug("Booting UG96\r\n");
+    m_reset.write(1);
+    wait_ms(200);
+    m_reset.write(0);
+    wait_ms(150);
+    m_powerkey.write(1);
+    wait_ms(150);
+    m_powerkey.write(0);
+    /* Because modem status is not available on STMOD+ connector,
+     * let's wait for Modem complete boot */
+    wait_ms(2300);
+#endif
+#if (MBED_CONF_STMOD_CELLULAR_TYPE == STMOD_BG96)
+    tr_debug("Booting BG96\r\n");
+    m_powerkey.write(1);
+    m_reset.write(1);
+    wait_ms(150);
+    m_powerkey.write(0);
+    m_reset.write(0);
+    wait_ms(100);
+    m_powerkey.write(1);
+    wait_ms(200);
+    m_powerkey.write(0);
+    wait_ms(5000);
+#endif
+
+    nsapi_error_t err = STMOD_CELLULAR_MODEM::soft_power_on();
     if (err != 0) {
         return err;
     }
-    m_powerkey.write(1);
-    wait_ms(250);
-    m_powerkey.write(0);
 
     // wait for RDY
     _at->lock();
     _at->set_at_timeout(5000);
     _at->set_stop_tag("RDY");
     bool rdy = _at->consume_to_stop_tag();
-    tr_debug("Modem %sready to receive AT commands", rdy?"":"NOT ");
     (void)rdy;
+
+    /*  Modem may send more bytes are RDY flag */
+    _at->flush();
+
+    /* Turn OFF ECHO before anything else */
```

Comment:
Is this required here since it is also done by the `CellularStateMachine` ?

Reply:
We want to turn it off as soon as possible, before Flow control commands, otherwise we're getting RX characters in AT fifo that seem to be not well parsed and may lead to error later on ... I've seen it more stable when doing so.